mkosi/btrfs.py

```diff
@@ -0,0 +1,46 @@
+from pathlib import Path
+import shutil
+
+from mkosi.config import MkosiConfig, ConfigFeature
+from mkosi.log import die
+from mkosi.run import run
+from mkosi.install import copy_path
+
+
+def btrfs_maybe_make_subvolume(config: MkosiConfig, path: Path, mode: int) -> None:
+    if config.use_subvolumes == ConfigFeature.enabled and not shutil.which("btrfs"):
+        die("Subvolumes requested but the btrfs command was not found")
+
+    if config.use_subvolumes != config.use_subvolumes.disabled:
+        result = run(["btrfs", "subvolume", "create", path], check=config.use_subvolumes == ConfigFeature.enabled).returncode
+    else:
+        result = 0
+
+    if result == 0:
+        path.chmod(mode)
+    else:
+        path.mkdir(mode)
+
+
+def btrfs_maybe_snapshot_subvolume(config: MkosiConfig, src: Path, dst: Path) -> None:
+    subvolume = (config.use_subvolumes == ConfigFeature.enabled or
+                 config.use_subvolumes == ConfigFeature.auto and shutil.which("btrfs") is not None)
+
+    if config.use_subvolumes == ConfigFeature.enabled and not shutil.which("btrfs"):
+        die("Subvolumes requested but the btrfs command was not found")
+
+    # Subvolumes always have inode 256 so we can use that to check if a directory is a subvolume.
+    if not subvolume or src.stat().st_ino != 256 or (dst.exists() and any(dst.iterdir())):
+        return copy_path(src, dst)
+
+    # btrfs can't snapshot to an existing directory so make sure the destination does not exist.
+    if dst.exists():
+        dst.rmdir()
+
+    # Only read-only subvolumes can be snapshotted so make the subvolume read-only while we do the snapshot.
```

Comment:
I'm not sure about this comment?
```
btrfs subvolume create test
touch test/foobar
btrfs subvolume snapshot test test2
```
works perfectly well for me

Reply:
Ah, I think I got confused, it's send receive that must happen read-only